x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportResetJobAction.java

```diff
@@ -0,0 +1,243 @@
+                JobState jobState = MlTasks.getJobState(job.getId(), tasks);
+                if (request.isSkipJobStateValidation() == false && jobState != JobState.CLOSED) {
+                    listener.onFailure(ExceptionsHelper.conflictStatusException(Messages.getMessage(Messages.REST_JOB_NOT_CLOSED_RESET)));
+                    return;
+                }
+                if (job.getBlocked().getReason() != Blocked.Reason.NONE && job.getBlocked().getReason() != Blocked.Reason.RESET) {
+                    listener.onFailure(ExceptionsHelper.conflictStatusException(
+                        "cannot reset job while it is blocked with [" + job.getBlocked().getReason() + "]"));
+                    return;
+                }
+
+                ParentTaskAssigningClient taskClient = new ParentTaskAssigningClient(client, taskId);
+                jobConfigProvider.updateJobBlockReason(job.getId(), new Blocked(Blocked.Reason.RESET, taskId), ActionListener.wrap(
+                    r -> resetJob(taskClient, (CancellableTask) task, request, listener),
+                    listener::onFailure
+                ));
+            },
+            listener::onFailure
+        );
+
+        ActionListener<TaskInfo> existingTaskListener = ActionListener.wrap(
+            existingTask -> {
+                if (existingTask == null) {
+                    jobConfigProvider.getJob(request.getJobId(), jobListener);
+                } else {
+                    waitExistingResetTaskToComplete(existingTask, request, listener);
+                }
+            },
+            listener::onFailure
+        );
+
+        findExistingResetTask(taskId, request.getJobId(), existingTaskListener);
+    }
+
+    private void findExistingResetTask(TaskId currentTaskId, String jobId, ActionListener<TaskInfo> listener) {
+        ListTasksRequest listTasksRequest = new ListTasksRequest();
+        listTasksRequest.setActions(ResetJobAction.NAME);
+        listTasksRequest.setDescriptions(MlTasks.JOB_TASK_ID_PREFIX + jobId);
+        listTasksRequest.setDetailed(true);
+        listTasksRequest.setNodes(currentTaskId.getNodeId());
+        executeAsyncWithOrigin(client, ML_ORIGIN, ListTasksAction.INSTANCE, listTasksRequest, ActionListener.wrap(
+            listTasksResponse -> {
+                Optional<TaskInfo> existingResetTask = listTasksResponse.getTasks().stream()
+                    .filter(taskInfo -> taskInfo.getTaskId().equals(currentTaskId) == false).findAny();
+                if (existingResetTask.isPresent()) {
+                    listener.onResponse(listTasksResponse.getTasks().get(0));
+                } else {
+                    listener.onResponse(null);
+                }
+            },
+            listener::onFailure
+        ));
+    }
+
+    private void waitExistingResetTaskToComplete(TaskInfo existingTask, ResetJobAction.Request request,
```

Comment:
I think this and the above list method can be adjusted to look at the current `blocked.task_id` and if the task doesn't exist, start another one and update the doc.